### DISCUSSION TITLE
Trigger single frame GPU Trace externally

### DISCUSSION THREAD

**Discussion by IsaacMarovitz** [KhronosGroup/MoltenVK#1859] [Q&A] [answered]
Hi there, I'm trying to debug one specific area in a game that struggles and could benefit from using a GPU Trace to diagnose the issue. The problem is when using `MVK_CONFIG_AUTO_GPU_CAPTURE_SCOPE=1`, the file sizes get too big too quickly to the point where Xcode won't open them. The specific frames I want to debug don't start until a couple of minutes into gameplay, and I only need a single frame sampled. Is there a way I can externally trigger a trace on a specific frame?

**Comment by billhollings** [comment]
The best way to do this is to run your app from Xcode, where you can enable GPU frame capture, and manually trigger the frame you want to capture at the appropriate time during your game run.

You do **NOT** need to create a build environment (files, code, build targets) within Xcode. Just create a Scheme, and in the Run section, select the executable to launch for the run, select _Debug Build Configuration_, and enable _Debug executable_. You can set any command line arguments and environmental variables your app needs, in the Arguments tab, and in the Options tab, select _GPU Frame Capture: Metal_.

If you encounter any issues, or in case I've forgotten something in my explanation, post any questions back here.

![Screenshot 2023-04-03 at 1 52 20 PM](https://user-images.githubusercontent.com/356402/229588408-ab144e51-a7e1-4d4a-9a90-5e9dfcb30cc8.jpg)

**Reply by IsaacMarovitz** [reply]
I ended up trying a similar method. The game I'm trying to diagnose is running under Wine, so I opted instead to launch the game with the appropriate environment variables and then attach the Xcode debugger to the process. The issue I ran into there, unfortunately, was that Xcode would get stuck on poll events after a few seconds, pausing the application and the GPU capture. The capture seems to take particularly long in the problem area I'm trying to diagnose due to poor performance, but without more Xcode stability I'm not sure how I can resolve this.

**Reply by andyoneal** [reply]
I've also run into the polling breakpoint issue and haven't figured out how to work around it besides racing to hit the "capture frame" button in between breaks. Would be nice to know what causes it.

**Reply by IsaacMarovitz** [reply]
<img width="272" alt="Screenshot 2023-04-09 at 14 53 13" src="https://user-images.githubusercontent.com/42140194/230791273-b1729d66-1ee7-4e72-8124-ffcfec615b8d.png">

Example of what Xcode gets stuck on, might be worth asking the Wine guys

**Reply by IsaacMarovitz** [reply, marked as answer]
For those wondering on updates, I've made a patch that creates a separate window from which you can trigger a capture.

[0001-Frame-Capture-Tool-Window.patch](https://github.com/KhronosGroup/MoltenVK/files/11408795/0001-Frame-Capture-Tool-Window.patch)
